File: leetcode/1-100/74-search-a-2d-matrix/bench/search_a_2d_matrix.c
/*
 * Benchmark workload — Search a 2D Matrix (LeetCode #74).
 * C mirror of bench/search_a_2d_matrix.kara. Flattened binary search over a
 * 100x100 matrix built ONCE as int64_t** (row pointers into per-row mallocs —
 * double-indirection, matching Kara's Vec[Vec[i64]] indexing, NOT a flat 1-D
 * array), then K=10,000,000 queries (targets k % 20000, ~half hit) folding each
 * hit/miss bit into a rolling polynomial hash. See ../README.md § Benchmarks.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

static int search_matrix(int64_t **m, int64_t rows, int64_t cols, int64_t target) {
    if (rows == 0) return 0;
    if (cols == 0) return 0;
    int64_t lo = 0, hi = rows * cols - 1;
    while (lo <= hi) {
        int64_t mid = lo + (hi - lo) / 2;
        int64_t v = m[mid / cols][mid % cols];
        if (v == target) return 1;
        else if (v < target) lo = mid + 1;
        else hi = mid - 1;
    }
    return 0;
}

int main(void) {
    const int64_t rows = 100, cols = 100;
    const int64_t total = 10000000, modulus = 1000000007;
    const int64_t range = 2 * rows * cols;

    int64_t **m = (int64_t **)malloc(sizeof(int64_t *) * (size_t)rows);
    for (int64_t i = 0; i < rows; i++) {
        m[i] = (int64_t *)malloc(sizeof(int64_t) * (size_t)cols);
        for (int64_t j = 0; j < cols; j++)
            m[i][j] = (i * cols + j) * 2;
    }

    int64_t acc = 0;
    for (int64_t k = 0; k < total; k++) {
        int64_t target = k % range;
        int64_t bit = search_matrix(m, rows, cols, target) ? 1 : 0;
        acc = (acc * 131 + bit) % modulus;
    }
    printf("%lld\n", (long long)acc);

    for (int64_t i = 0; i < rows; i++) free(m[i]);
    free(m);
    return 0;
}

File: leetcode/1-100/74-search-a-2d-matrix/bench/search_a_2d_matrix.rs
// Benchmark workload — Search a 2D Matrix (LeetCode #74).
// Rust mirror of bench/search_a_2d_matrix.kara. Flattened binary search over a
// 100×100 matrix built ONCE as a Vec<Vec<i64>> (double-indirection, matching
// Kāra's Vec[Vec[i64]]), then K=10_000_000 queries (targets k % 20000, ~half hit)
// folding each hit/miss bit into a rolling polynomial hash. The search is the
// measured work. See ../README.md § Benchmarks.

fn search_matrix(m: &Vec<Vec<i64>>, target: i64) -> bool {
    let rows = m.len();
    if rows == 0 {
        return false;
    }
    let cols = m[0].len();
    if cols == 0 {
        return false;
    }
    let mut lo: i64 = 0;
    let mut hi: i64 = (rows * cols) as i64 - 1;
    while lo <= hi {
        let mid = lo + (hi - lo) / 2;
        let v = m[(mid / cols as i64) as usize][(mid % cols as i64) as usize];
        if v == target {
            return true;
        } else if v < target {
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    false
}

fn main() {
    const ROWS: i64 = 100;
    const COLS: i64 = 100;
    const TOTAL: i64 = 10_000_000;
    const MODULUS: i64 = 1_000_000_007;
    let range = 2 * ROWS * COLS;

    let mut m: Vec<Vec<i64>> = Vec::new();
    for i in 0..ROWS {
        let mut row: Vec<i64> = Vec::new();
        for j in 0..COLS {
            row.push((i * COLS + j) * 2);
        }
        m.push(row);
    }

    let mut acc: i64 = 0;
    for k in 0..TOTAL {
        let target = k % range;
        let bit = if search_matrix(&m, target) { 1 } else { 0 };
        acc = (acc * 131 + bit) % MODULUS;
    }
    println!("{}", acc);
}
